ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -294,3 +324,116 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+            in Just $ WhereBinding ids l
+        findBindingIds _ = Nothing
+
+        -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
+        findIdFromPat :: Pat GhcTc -> Maybe Id
+        findIdFromPat (VarPat _ (L _ id)) = Just id
+        findIdFromPat _                   = Nothing
+
+        findSigIds (L _ (TypeSig _ names _)) = map unLoc names
+        findSigIds _                         = []
+
+-- | Provide code lens for where bindings.
+whereClauseCodeLens :: PluginMethodHandler IdeState TextDocumentCodeLens
+whereClauseCodeLens state plId CodeLensParams{..} = do
+    enabled <- usePropertyLsp #whereLensOn plId properties
+    if not enabled then pure $ pure $ List [] else pluginResponse $ do
+        nfp <- getNormalizedFilePath plId uri
+        tmr <- handleMaybeM "Unable to typechecking"
+            $ liftIO
+            $ runAction "codeLens.local.TypeCheck" state
+            $ use TypeCheck nfp
+        (hscEnv -> hsc) <- handleMaybeM "Unable to get GhcSession"
+            $ liftIO
+            $ runAction "codeLens.local.GhcSession" state
+            $ use GhcSession nfp
+        let tcGblEnv = tmrTypechecked tmr
+            rdrEnv = tcg_rdr_env tcGblEnv
+            typeCheckedSource = tcg_binds tcGblEnv
+
+            wheres = findWhereQ typeCheckedSource
+            localBindings = mapMaybe findBindingsQ wheres
+
+            -- | Note there may multi ids for one binding
+            bindingToLenses ids span = case srcSpanToRange span of
+                Nothing -> pure []
+                Just range -> forM ids $ \id -> do
+                    (_, fromMaybe [] -> sig) <- liftIO
+                        $ initTcWithGbl hsc tcGblEnv ghostSpan
+                        $ bindToSig id hsc rdrEnv
+                    pure $ generateWhereLens plId range (T.pack sig)
+
+        lenses <- concat <$> sequence
+            [ bindingToLenses idsWithoutSig bindingLoc
+            | WhereBindings{..} <- localBindings
+            , let sigSpans = getSrcSpan <$> existedSigNames
+            , WhereBinding{..} <- bindings
+            , let idsWithoutSig = filter (\x -> getSrcSpan (idName x) `notElem` sigSpans) bindingId
+            ]
+
+        pure $ List lenses
+        where
+            uri = _textDocument ^. L.uri
+
+            generateWhereLens :: PluginId -> Range -> T.Text -> CodeLens
+            generateWhereLens plId range title =
+                let cmd = mkLspCommand plId typeLensCommandId title (Just [toJSON (makeEdit range title)])
+                in  CodeLens range (Just cmd) Nothing
+
+            makeEdit :: Range -> T.Text -> WorkspaceEdit
+            makeEdit range text =
+                let startPos = range ^. L.start
+                    insertChar = startPos ^. L.character
+                    insertRange = Range startPos startPos
+                in WorkspaceEdit
+                    (pure [(uri, List [TextEdit insertRange (text <> "\n" <> T.replicate (fromIntegral insertChar) " ")])])
```

Comment:
huh, do we really have to do this? insert spaces on the next line? I'm surprised there isn't something to handle that, and this seems to have the risk of breaking people who use tabs. worth a comment

Reply:
There's `InsertTextMode` but only for completions :roll_eyes:

Reply:
Insert space to align the signature and its definition.